src/lib/uri/uri.c

```diff
@@ -9,9 +9,111 @@
 
 #define XSTRNDUP(s, n)  (s != NULL ? xstrndup(s, n) : NULL)
 
+/**
+ * Found URI option by it's @a name in @a uri structure.
+ */
+static struct uri_query_param *
+uri_query_param_by_name(const struct uri *uri, const char *name)
+{
+        for (int i = 0; i < uri->option_count; i++) {
+                if (strcmp(uri->options[i].name, name) == 0)
+                        return &uri->options[i];
+        }
+        return NULL;
+}
+
+/**
+ * Add new @a value to URI @a option.
+ */
+static void
+uri_query_param_add_value(struct uri_query_param *option, const char *value)
+{
+        size_t size = (option->value_count + 1) * sizeof(char *);
+        option->values = xrealloc(option->values, size);
+        option->values[option->value_count++] = xstrdup(value);
+}
+
+/**
+ * Destroy URI @a option and free all associated resourses.
+ */
+static void
+uri_query_param_destroy(struct uri_query_param *option)
+{
+        for (int i = 0; i < option->value_count; i++)
+                free(option->values[i]);
+        free(option->values);
+        free(option->name);
+        TRASH(option);
+}
+
+/**
+ * Create new URI @a option, with given @a name.
+ */
+static void
+uri_query_param_create(struct uri_query_param *option, const char *name)
+{
+        memset(option, 0, sizeof(struct uri_query_param));
+        option->name = xstrdup(name);
+}
+
+/**
+ * Add new option to @uri structure with given @a name.
+ */
+static struct uri_query_param *
+uri_add_query_param(struct uri *uri, const char *name)
+{
+      size_t size = (uri->option_count + 1) * sizeof(struct uri_query_param);
+      uri->options = xrealloc(uri->options, size);
+      struct uri_query_param *option = &uri->options[uri->option_count++];
+      uri_query_param_create(option, name);
+      return option;
+}
+
+/**
+ * Destroy @a uri options and all associated resourses.
+ */
+static void
+uri_query_params_destroy(struct uri *uri)
+{
+        for (int i = 0; i < uri->option_count; i++)
+                uri_query_param_destroy(&uri->options[i]);
+        free(uri->options);
+}
+
+/**
+ * Create options for @a uri from @a query string. Expected @a query
+ * format is a string which contains options separated by '&'. For
+ * example: "backlog=10&transport=tls". Also @a query can contains
+ * several values for one option separated by '&'. For example:
+ * "backlog=10&backlog=30".
+ */
+static void
+uri_query_params_create(struct uri *uri, const char *query)
```

Comment:
A nit pick about naming: `obj_do` - is supposed to do something with `obj` so I'd call functions a bit differently:

`uri_query_params_create` -> `uri_create_query_params`
`uri_query_params_destroy` -> `uri_destroy_query_params`
`uri_query_param_by_name` -> `uri_find_query_param` (to match `uri_add_query_param`).